host/xtest/regression_9000.c

```diff
@@ -0,0 +1,174 @@
+ */
+#include <fcntl.h>
+#include "include/arm_ffa_user.h"
+#include "include/ffa.h"
+#include <stdio.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+#include "xtest_helpers.h"
+#include "xtest_test.h"
+
+#define FFA_DRIVER_FS_PATH	"/sys/kernel/debug/arm_ffa_user"
+#define SPMC_TEST_OK 0xaa
+
+enum sp_tests {
+	EP_TEST_SP,
+	EP_TEST_SP_COMMUNICATION,
+};
+
+static int fp = -1;
+
+const char test_endpoint1_uuid[] = "5c9edbc3-7b3a-4367-9f83-7c191ae86a37";
+const char test_endpoint2_uuid[] = "7817164c-c40c-4d1a-867a-9bb2278cf41a";
+const char test_endpoint3_uuid[] = "23eb0100-e32a-4497-9052-2f11e584afa6";
+
+struct ffa_ioctl_ep_desc test_endpoint1 = {
+		.uuid_ptr = (uint64_t) test_endpoint1_uuid,
+		.id = 0
+};
+
+struct ffa_ioctl_ep_desc test_endpoint2 = {
+		.uuid_ptr = (uint64_t) test_endpoint2_uuid,
+		.id = 0
+};
+
+struct ffa_ioctl_ep_desc test_endpoint3 = {
+		.uuid_ptr = (uint64_t) test_endpoint3_uuid,
+		.id = 0
+};
+
+static void close_debugfs(void)
+{
+	int err;
+	if (fp >= 0) {
+		err = close(fp);
+		if (err < 0) {
+			fprintf(stderr, "Error: Could not close the FF-A driver\n");
+		}
+	}
+}
+
+static bool open_debugfs(void)
+{
+	if (fp < 0) {
+		fp = open(FFA_DRIVER_FS_PATH, O_RDWR);
+		if (fp < 0) {
+			fprintf(stderr, "Error: Could not open the FF-A driver\n");
+			return false;
+		}
+	}
+	return true;
+}
+
+static bool init(ADBG_Case_t *c)
+{
+	if (!open_debugfs()) {
+		Do_ADBG_Log("Failed to open debugFs");
+		return false;
+	}
+
+	return true;
+}
+
+static bool init_sp_xtest(ADBG_Case_t *c)
+{
+
+	if (!init(c)) {
+		Do_ADBG_Log("Failed to load kernel module");
+		return false;
+	}
+	return true;
+}
+
+
+static bool start_sp_test(uint16_t endpoint, enum sp_tests test,
+			  struct ffa_ioctl_msg_args *args)
+{
+	args->dst_id = endpoint;
+	args->args[0] = (uint32_t)test;
+	return (ioctl(fp, FFA_IOC_MSG_SEND, args) >= 0);
+}
+
+static uint16_t get_endpoint_id(uint64_t endp)
+{
+	struct ffa_ioctl_ep_desc sid = {0};
+
+	sid.uuid_ptr = endp;
+
+	/* Get ID of destination SP based on UUID */
+	if(ioctl(fp, FFA_IOC_GET_PART_ID, &sid))
+		return 0xffff;
+
+	return sid.id;
+}
+
+static bool check_ok(struct ffa_ioctl_msg_args *args)
+{
+	if (args->args[0] != SPMC_TEST_OK) {
+		Do_ADBG_Log("Received incorrect value %x", args->args[0]);
+		return false;
+	}
+
+	return true;
+}
+
+static void xtest_tee_test_9001(ADBG_Case_t *c)
+{
+	struct ffa_ioctl_msg_args args = { 0 };
+	uint16_t endpoint1_id = 0;
+	uint16_t endpoint2_id = 0;
+
+	if (!init_sp_xtest(c)) {
+		Do_ADBG_Log("Failed to initialise test, skipping SP test");
+		goto out;
+	}
+
+	endpoint1_id = get_endpoint_id(test_endpoint1.uuid_ptr);
+	Do_ADBG_Log("Endpoint1 id %x",endpoint1_id);
+	if (endpoint1_id == 0xffff) {
+		Do_ADBG_Log("Could not contact xtest_1 sp, skipping SP test");
+		Do_ADBG_Log("Add xtest_1 sp to the image to enable tests");
+		goto out;
+	}
+
+	memset(&args, 0, sizeof(args));
+	ADBG_EXPECT_TRUE(c, start_sp_test(endpoint1_id, EP_TEST_SP, &args));
```

Comment:
If either of these fails, does it make any sense to continue?

Reply:
Ack